### PR TITLE
miniupnpd-nftables: init at 2.3.3

### LINKED ISSUE #246345: Package request: miniupnpd-nftables
We already have [miniupnpd](https://github.com/miniupnp/miniupnp) in [nixpkgs](https://github.com/NixOS/nixpkgs/tree/0d117d7ad5d590991d23ccc7bd88c4e627cccf17/pkgs/tools/networking/miniupnpd), but only with iptables support. It would be nice to have the nftables version available as well.

### PR DESCRIPTION
**Pull request by chayleaf**
###### Description of changes

This allows using miniupnpd with the nftables firewall. Closes #246345

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/miniupnpd/default.nix [hunk 1]
```diff
@@ -15,22 +16,40 @@ stdenv.mkDerivation rec {
...
+  buildInputs = [ iptables-legacy libuuid openssl ]
+    ++ lib.optionals (firewall == "nftables") [ libmnl libnftnl ];
   nativeBuildInputs= [ pkg-config makeWrapper ];
 
 
   # ./configure is not a standard configure file, errors with:
   # Option not recognized : --prefix=
   dontAddPrefix = true;
 
+  FW = firewall;
```
Comment (RaitoBezarius) [inline pkgs/tools/networking/miniupnpd/default.nix:28]: Please prefix all environment variables `env.FW = ...`, `env.IPV6 = ...` or just do `env = { FW = ...; … }`
(this is a way to separate options from env vars we pass to the drv environment)

File: pkgs/tools/networking/miniupnpd/default.nix [hunk 2]
```diff
@@ -15,22 +16,40 @@ stdenv.mkDerivation rec {
...
+  # allow using various config options
+  IPV6 = true;
```
Comment (RaitoBezarius) [inline pkgs/tools/networking/miniupnpd/default.nix:30]: is there any reason those are not makefile options?

**Review by RaitoBezarius — APPROVED** [review]
LGTM modulo `env`stuff, untested yet

File: pkgs/tools/networking/miniupnpd/default.nix (hunk 1 above)
Comment (chayleaf) [reply pkgs/tools/networking/miniupnpd/default.nix]: I use configureFlags now

File: pkgs/tools/networking/miniupnpd/default.nix (hunk 2 above)
Comment (chayleaf) [reply pkgs/tools/networking/miniupnpd/default.nix]: I remember there being some problem with configureFlags not having any effect, but it works now so I guess I just made a mistake back then

**Comment by chayleaf** [comment]
Good news is I've added a test. Bad news is it doesn't pass. Oh well.

**Comment by justinas** [comment]
Perhaps my WIP will help. https://github.com/justinas/nixpkgs/commit/68ad106b675492092d32744511a2cf41220edc9e

I chose a different approach and decided to forgo the table initialization scripts that are bundled with miniupnpd. In my eyes, these are problematic in conjunction with nftables-based NixOS firewall, since when you have chains for the same hook in different tables:
> The ordering in the example will be undefined, but both chains will be traversed (unless for example the packet gets dropped in the first chain seen). https://unix.stackexchange.com/questions/607358/packet-processing-order-in-nftables

Instead, I reuse the `nixos-fw` table, and add miniupnpd chains there. Mostly I was waiting for https://github.com/NixOS/nixpkgs/pull/207758 to merge first, so we can declare this chain declaratively as well as make miniupnpd compatible with `filterForward`.

**Comment by chayleaf** [comment]
The problem is there's no way to restart a systemd service upon another service's reload, so if nftables is reloaded miniupnpd won't know about it and won't re-initialize the table. This problem can be ignored, but a potential solution is forcing nftables *restart* (rather than reload) if miniupnpd is enabled.

**Comment by justinas** [comment]
> However, in reality issues would arise in conjunction with stuff like `filterForward`

I was actually wrong here, as NixOS' nftables firewall already does the correct thing and allows forwarding for anything DNAT'ed

https://github.com/NixOS/nixpkgs/blob/d1d57f39aa1f41d6c45ec256a81ee3a8f05acb7e/nixos/modules/services/networking/firewall-nftables.nix#L164

---

Now that #207758 has merged, it should be perfectly viable to declare miniupnpd's tables declaratively and avoid the scripts altogether:

```nix
{
  # This is essentially a no-op table - we don't need this because
  # NixOS will allow forwarding anything DNATed by default.
  # Omitting the forward chain and its "drop" policy here avoids
  # two forward chains (miniupnpd and nixos-fw) stepping on each others toes.
  networking.nftables.tables.miniupnpd = {
    family = "inet";
    content = ''
      chain miniupnpd {}
    '';
  };

  networking.nftables.tables.miniupnpd-nat = {
    family = "inet";
    content = ''
      chain prerouting_miniupnpd {
        type nat hook prerouting priority dstnat;
      }
      chain postrouting_miniupnpd {
        type nat hook postrouting priority srcnat;
      }
    '';
  };
}
```

**Comment by chayleaf** [comment]
thanks, I'll get to it later

**Review by marcusramberg — APPROVED** [review]
the module change lgtm, and pr review built ok on x86 and aarch64.

**Review by justinas — APPROVED** [review]

**Pull request merged**